cpop_bot/new_members_handler.py

```diff
@@ -97,6 +100,11 @@ async def handle_captcha_callback(self, callback: types.CallbackQuery):
                 from_user_id,
                 types.ChatPermissions(True, True, True, True,
                                       True, True, True, True))
+            await self.__delay_delete_message(DELAY_DELETE_CONGRATS, congrats)
```

Comment:
i guess we should remove the pending user from the list here as well then? like this
```
        pending_user = self.__get_pending_user(message_id, chat_id)
        if pending_user.user_id != from_user_id:
            await self.bot.answer_callback_query(
                callback.id,
                text="We are waiting for {0} to click".format(
                    (await callback.message.chat.get_member(pending_user.user_id))
                    .user.first_name),
                show_alert=False)
        else:
            congrats = await self.bot.edit_message_text(
                chat_id=chat_id, message_id=message_id,
                text="Congrats {0}, captcha passed".format(
                    self.__get_mention(callback.from_user)),
                parse_mode=ParseMode.HTML, reply_markup=None)
            await self.bot.restrict_chat_member(
                chat_id,
                from_user_id,
                types.ChatPermissions(True, True, True, True,
                                      True, True, True, True))
            self.pending_users.remove(pending_user)
            await self.__delay_delete_message(DELAY_DELETE_CONGRATS, congrats)
```

right now it gets removed from the list in **__purge_old_captchas(self)** method, so if we remove it here, then **__purge_old_captchas(self)** will only be responsible for removing messages with captcha buttons

Reply:
Yeah, I thought there could be problem like this. I didn't fully understand those codes relating to that pending_users list.

Pushed new commit. It runs fine but idk how to check the list, I think it's fine now.

Reply:
`idk how to check the list`

you want to check the list to make sure the user was actually removed? You can temporary add `logger.debug(self.pending_users)` in places where you make a change to the list and see the output in the console

Reply:
> you want to check the list to make sure the user was actually removed?

yes

> logger.debug(self.pending_users)

this won't work logger only accept type `str`, it should be converted to string at first but idk how.

maybe just leave it

btw ``` `code` ``` backtick is usually used for quote source code to show it in monospace style, to quote normal sentences, it's better to use ```> quotes```, just `>` at the beginning of the line

Reply:
sorry, didn't notice the message

> this won't work logger only accept type str, it should be converted to string at first but idk how.
yeah, i forgot, you'd need to make a string representation for PendingUser class, but it's pretty simple https://stackoverflow.com/questions/4932438/how-to-create-a-custom-string-representation-for-a-class-object

so this should work
```
class PendingUser:
    def __init__(self, user_id, captcha_message: types.Message, timestamp):
        self.user_id = user_id
        self.captcha_message = captcha_message
        self.timestamp = timestamp

    def __str__(self):
        return "user_id: " + self.user_id + ", captcha_message: " \
               + self.captcha_message.text + ", timestamp: " + self.timestamp
```

and now you just need to convert a **list** of **PendingUser** objects to a **string**, apparently for this you can use a **join** method, so i presume this should do the trick
```
        logger.debug(''.join(str(user) for user in self.pending_users))

```

Reply:
btw if you still want to test with the logger, you can actually add a commit with `__str__` method  added for **PendingUser** class, since it's a pretty useful method

Reply:
Just tried, it works. but I think I may not do it properly, so I won't push commit for this. (maybe it's better to just output user's name and id, and the debug info may be printed at multiple places in the code, that sounds a bit hard to deal with.)

Reply:
Will mark this as resolved and merge this, this PR works properly as expected from the debug output I tried.